stonecrop/docs/FIELD_TRIGGERS.md

```diff
@@ -0,0 +1,894 @@
+- [Action Function Interface](#action-function-interface)
+- [Execution Model](#execution-model)
+- [Pattern Matching](#pattern-matching)
+- [Advanced Configuration](#advanced-configuration)
+- [Error Handling](#error-handling)
+- [Automatic Rollback](#automatic-rollback)
+- [Best Practices](#best-practices)
+- [Schema Composability](#schema-composability)
+- [API Reference](#api-reference)
+- [Migration Guide](#migration-guide)
+
+## Architecture
+
+The `FieldTriggerEngine` uses a **singleton pattern**:
+
+```typescript
+// Creating new instances returns the same singleton
+const engine1 = new FieldTriggerEngine()
+const engine2 = new FieldTriggerEngine()
+// engine1 === engine2 === FieldTriggerEngine._root
+
+// Alternatively, use the helper function
+const engine = getGlobalTriggerEngine()
+```
+
+This ensures:
+- Consistent action registry across the application
+- Shared doctype action mappings
+- Single source of truth for field trigger configuration
+
+## Basic Usage
+
+### 1. Register Actions (Optional)
+
+You can optionally register global action functions for reuse across doctypes:
+
+```typescript
+import { registerGlobalAction } from '@stonecrop/stonecrop'
+
+// Register a global action function
+registerGlobalAction('validateEmailPrimary', (context) => {
+  console.log('Email primary validation triggered:', context.afterValue)
+  // Validation logic here
+})
+
+registerGlobalAction('updateTimestamp', async (context) => {
+  // Async action example
+  await updateModifiedTime(context.doctype, context.recordId)
+})
+```
+
+### 2. Define Actions and Field Triggers in Doctype
+
+Add both regular actions and field triggers to your doctype's actions Map:
+
+```typescript
+import { DoctypeMeta } from '@stonecrop/stonecrop'
+import { Map } from 'immutable'
+
+const actions = Map({
+  // Regular actions (triggered manually)
+  'save': ['updateTimestamp', 'validateData'],
+  'delete': ['confirmDelete', 'cleanupReferences'],
+
+  // Field triggers (triggered automatically on field changes)
+  'status': ['updateTimestamp'],                     // Simple field trigger
+  'emailAddress.*.is_primary': ['validateEmailPrimary'], // Wildcard pattern
+  'profile.name': ['validateName', 'updateTimestamp'], // Multiple actions
+  'title': ['(context) => console.log("Title changed:", context.afterValue)'], // Inline function
+})
+
+const doctype = new DoctypeMeta(
+  'Contact',
+  schema,
+  workflow,
+  actions,
+  component
+)
+```
+
+### 3. HST Operations Trigger Actions
+
+Field triggers are automatically executed when HST paths are modified:
+
+```typescript
+const store = stonecrop.getStore()
+
+// This will trigger the 'validateEmailPrimary' action
+store.set('Contact.contact-1.emailAddress.0.is_primary', true)
+
+// This will trigger the 'updateTimestamp' action
+store.set('Contact.contact-1.status', 'completed')
+```
+
+## Action Function Interface
+
+Action functions are simple, focused functions that receive a `FieldChangeContext` object:
```

Comment:
Worth noting here that there are separate schema driven mask and format functions?